### PR TITLE
Factorize mod schema

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
There is a lot of code duplication for each changes that need a full sync. I factorize it.

Furthermore, I discovered that the app did not ask the user to confirm whether they want to confirm something when they need to confirm they want a full sync. I changed it so that it asks to confirm both.

Actually, asking for confirmation before asking the user to enter a name for a field would make sens too.

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java [hunk 1]
```diff
@@ -609,5 +612,46 @@ protected boolean showedActivityFailedScreen(Bundle savedInstanceState) {
...
+            ConfirmationDialog d = new ConfirmationDialog();
+            d.setArgs(getResources().getString(R.string.full_sync_confirmation));
+            d.setConfirm(
+                    () -> {
+                        try {
+                            getCol().modSchemaNoCheck();
+                            fn.run();
+                        } catch (ConfirmModSchemaException cmse) {
+                            // This should never be reached as we just forced modSchema
+                            throw new RuntimeException(cmse);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java:646]: It is very occasionally - more logging in this area would be appreciated as I don't see how it's possible

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java [hunk 2]
```diff
@@ -609,5 +612,46 @@ protected boolean showedActivityFailedScreen(Bundle savedInstanceState) {
...
+
+    protected interface SchemaModifier {
+        void run() throws ConfirmModSchemaException;
+    }
+
+
+    /**
+     * @param fn A function to execute if schema can be modified
+     */
+    protected void executeSchemaModified(@NonNull SchemaModifier fn) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java:624]: Implementer's choice: This might be more readable as a builder

**Review by david-allison-1 — COMMENTED** [review]
At a cursory glance, looks good, I'll need to go through it with a fine toothcomb later

Unit tests would be nice, since they're flagged by our coverage tooling

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java (hunk 1 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java:646]: Timber log, or send exception to server ?

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java (hunk 2 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java:624]: I don't see the interest of the builder when there are so few members and one is mandatory. I'm not convinced here

**Comment by Arthur-Milchior** [comment]
I'll take a look to figure out how to test activities when I'm done with templates

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java (hunk 1 above)
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java:646]: Timber logs - this currently crashes, posting logcat and the logs will be usable in debugging

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.java [hunk 3]
```diff
@@ -290,26 +268,7 @@ private void renameFieldDialog() {
...
-                        ConfirmationDialog c = new ConfirmationDialog();
-                        c.setArgs(getResources().getString(R.string.full_sync_confirmation));
-                        Runnable confirm = () -> {
-                            mCol.modSchemaNoCheck();
-                            try {
-                                renameField();
-                            } catch (ConfirmModSchemaException e1) {
-                                //This should never be thrown
-                            }
-                            dismissContextMenu();
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.java:307]: dismissContextMenu is no longer called

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.java [hunk 4]
```diff
@@ -343,31 +302,12 @@ private void repositionFieldDialog() {
...
-                                    } catch (JSONException e1) {
-                                        throw new RuntimeException(e1);
-                                    }
-                                };
-                                c.setConfirm(confirm);
-                                c.setCancel(mConfirmDialogCancel);
-                                ModelFieldEditor.this.showDialogFragment(c);
-                            }
+                            executeSchemaModified(() -> {
+                                TaskManager.launchCollectionTask(new CollectionTask.RepositionField(mMod,
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.java:306]: unsure about this one

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
I haven't had the time to look through all of these, I'm not confident that this will be bug-free (or ever was in the first place), and testing, or more analysis may be appropriate for some of the harder cases.

**Comment by Arthur-Milchior** [comment]
This was clearly not meant to be closed. I understand it need more tests, and I need to understand how to test activities. But avoiding code duplication is still very relevant

**Comment by Arthur-Milchior** [comment]
@david-allison-1 If I removed the cases where you have doubt and did a PR containing only the cases that are easy to check for NF, would you accept it?

**Comment by david-allison-1** [comment]
Definitely

**Comment by Arthur-Milchior** [comment]
I removed the elements where you added comment. We should first commit #8390 and #8391

**Comment by david-allison** [comment]
@Arthur-Milchior Would you be happy to close this? 

Now we're in Kotlin this can be done in a much nicer manner.

**Pull request closed without merging**